driver-core/src/main/java/com/datastax/driver/core/RollingCount.java

```diff
@@ -0,0 +1,121 @@
+
+/**
+ * A "rolling" count over a 1-minute sliding window.
+ */
+class RollingCount {
+    // Divide the minute into 5-second intervals
+    private static final long INTERVAL_SIZE = TimeUnit.SECONDS.toNanos(5);
+
+    // A circular buffer containing the counts over the 12 previous intervals. Their sum is the count we're looking for.
+    // If we're at t = 61s this would span [0,60]
+    private final AtomicLongArray previousIntervals = new AtomicLongArray(12);
+    // The interval we're currently recording events for. It hasn't completed yet, so it's not included in the count.
+    // If we're at t = 61s this would span [60,65]
+    private final LongAdderV8 currentInterval = new LongAdderV8();
+    // Other mutable state, grouped in an object for atomic updates
+    private final AtomicReference<State> state;
+    private final com.codahale.metrics.Clock clock;
+
+    RollingCount(com.codahale.metrics.Clock clock) {
+        this.state = new AtomicReference<State>(new State(clock.getTick()));
+        this.clock = clock;
+    }
+
+    RollingCount() {
+        this(com.codahale.metrics.Clock.defaultClock());
+    }
+
+    void increment() {
+        add(1);
+    }
+
+    void add(long amount) {
+        tickIfNecessary();
+        currentInterval.add(amount);
+    }
+
+    long count() {
+        tickIfNecessary();
+        return state.get().totalCount;
+    }
+
+    private void tickIfNecessary() {
+        State oldState = state.get();
+        long newTick = clock.getTick();
+        long age = newTick - oldState.lastTick;
+        if (age > INTERVAL_SIZE) {
+            long currentCount = currentInterval.longValue();
+
+            long newIntervalStartTick = newTick - age % INTERVAL_SIZE;
+            long elapsedIntervals = Math.min(age / INTERVAL_SIZE, 12);
+            int newOffset = (int) ((oldState.offset + elapsedIntervals) % 12);
+
+            long newTotal;
+            if (elapsedIntervals == 12) {
+                // We wrapped around the circular buffer, all our values are stale
+                // Don't mutate previousIntervals yet because this part of the code is still multi-threaded.
+                newTotal = 0;
+            } else {
+                // Add the current interval that just completed
+                newTotal = oldState.totalCount + currentCount;
+                // Subtract all elapsed intervals: they're either idle ones, or the one at the old offset that we're
+                // about to replace
+                for (int i = 1; i <= elapsedIntervals; i++) {
+                    newTotal -= previousIntervals.get((newOffset + 12 - i) % 12);
+                }
+            }
+
+            State newState = new State(newIntervalStartTick, newOffset, newTotal);
+            if (state.compareAndSet(oldState, newState)) {
+                // Only one thread gets here, so we can now:
+                // - reset the current count (don't use reset because other threads might already have started updating
```

Comment:
I'm glad you documented this, when I first saw that `currentInterval` was separate from `state` my first thought was this possibly being a problem (state being updated and currentInterval being cleared out, when it may have been incremented by another thread).  Decrementing by the value captured in the beginning instead of clearing out will protect from counts being missed by this scenario. 👍